[GeoMechanicsApplication] Activate and improve 1D consolidation test

avdg81 opened the pull request:
**📝 Description**

It appeared that the 1D consolidation test was not part of any test suite, and hence it was never run automatically. Now it is part of the `nightly` test suite. At the same time, the test case needed to be improved:

- The analytical solution that is used for comparing the numerical solution of the relative water pressure field had some quirks, which became apparent after plotting it. We now follow more strictly the procedure put forward by Verruijt.
- A second test case has been added, which uses the C++ workflow for settlement analysis. It uses the exact same input files as the existing test case that uses the Python workflow. Also the assertions that are checked are identical for both test cases.

Other changes include:
- Added a helper function to read nodal positions from a `.post.msh` file.
- Properly calculate the total displacements. Removed a few redundant data members/local variables that kept a copy of the previous total displacements, since we can directly use the values stored at the nodes.

avdg81 committed:
Revived the 1D consolidation test

More cleanup work needs to be done.

avdg81 committed:
Renamed a Python test script

avdg81 committed:
Some minor cleanup

- Removed a few unused `import` statements.
- Removed redundant `setUp` and `tearDown` methods.
- Corrected a loop (forgot to `enumerate` the items of a list).

avdg81 committed:
Fixed failing integration test

The fix has been copied from another branch.

avdg81 committed:
The C++ settlement route also resets (total) displacements as needed

Also modified a unit test, to account for resetting the total displacements.

avdg81 committed:
Renamed the directory of the 1D consolidation test

Its name now matches the name of the Python test script.

avdg81 committed:
Corrected the C++ workflow for settlement analysis

With these corrections, we're able to run the 1D consolidation test using the C++ workflow for settlement analysis. The changes include:
- When no 'max_delta_time_factor' has been specified, assume a value of 1000.0 (just like the Python workflow does).
- Let the first stage of the 1D consolidation test also use solution_type 'dynamic' (to avoid an unsupported combination of solution_type and scheme_type).
- Refactored the Python script for running a settlement analysis. With these changes, we should be able to reuse it from the tests.
- Added support for the Newmark dynamic U-Pw scheme by the C++ workflow.

avdg81 committed:
Reuse script `run_geo_settlement.py` for the settlement workflow test

avdg81 committed:
Also run the 1D consolidation test using the C++ workflow

So far, it only checks the exit status. We should also check the relative water pressures as well as the degree of consolidation with the analytical solutions. This still needs to be done!

avdg81 committed:
Removed parameter "output_frequency" to fix several deprecation warnings

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/activate-1d-consolidation-test

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/activate-1d-consolidation-test

avdg81 committed:
Defined a common test base class for 1D consolidation

Also clarified some comments.

avdg81 committed:
Check the relative water pressures when using the C++ workflow for settlement

Also added a helper function that reads coordinates from a `.post.msh` file.

avdg81 committed:
Further improved the 1D consolidation test

- Defined two lists of node IDs as attributes of the test's base class.
- Calculate the degree of consolidation using the analytical formulation just once for all top nodes.

avdg81 committed:
Renamed a method

avdg81 committed:
Extracted a method for checking RMSE values

avdg81 committed:
Extracted a method for getting the y coordinates of a selection of nodes

avdg81 committed:
Further improved the 1D consolidation tests

- Renamed a few local variables
- Construct new lists using `append` rather than using indexes into lists filled with `None` values.

avdg81 committed:
Further improved the 1D consolidation tests

- Extracted methods to calculate the analytical solution, to retrieve the numerical solution, and to calculate the RMSE value of the differences between those solutions.
- The sample height is now an attribute of the test's base class.

avdg81 committed:
Further improved the 1D consolidation tests

- Renamed a few local variables.
- Reused the method that calculates the RMSE value of the differences between two lists of values.
- Create `GeoMechanicsAnalysis` instances on the fly rather than upfront.
- Don't use `enumerate` when the list index is no longer needed.

avdg81 committed:
Extracted a method that checks the degree of consolidation

avdg81 committed:
Reuse helper function `run_stages`

avdg81 committed:
Removed unused `import` statements and added a explaining comment

avdg81 committed:
Removed unnecessary copies of previous total displacement field

Now we directly add the incremental displacements stored at the nodes to the total displacements stored at the nodes. We've made sure that the incremental displacements are calculated before attempting to update the total displacements.

avdg81 committed:
Reformatted the code using Clang-format

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_3D_law.cpp:
These changes are unrelated to the aim of this PR, because they are the result of running Clang-format.

avdg81 commented on applications/GeoMechanicsApplication/custom_workflows/time_loop_executor.hpp:
```diff
@@ -70,8 +69,8 @@ class TimeLoopExecutor : public TimeLoopExecutorInterface
             NewEndState = RunCycleLoop(NewEndState);
-            mStrategyWrapper->AccumulateTotalDisplacementField();
             mStrategyWrapper->ComputeIncrementalDisplacementField();
+            mStrategyWrapper->AccumulateTotalDisplacementField();
```
Note that we can only update the total displacements **after** we have calculated the incremental displacements.

markelov208 commented on applications/GeoMechanicsApplication/custom_workflows/dgeosettlement.cpp:
```diff
@@ -72,7 +72,10 @@ double GetIncreaseFactorFrom(const Parameters& rProjectParameters)
 double GetMaxDeltaTimeFactorFrom(const Parameters& rProjectParameters)
 {
-    return rProjectParameters["solver_settings"]["time_stepping"]["max_delta_time_factor"].GetDouble();
+    // The provided default value here must match the one found in GeoMechanicsAnalysis.__init__
```
This comment is about 1000, right?

avdg81 replied:
Yes, that is correct. So when `"max_delta_time_factor"` has not been given by the user.

avdg81 commented on applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py:
```diff
@@ -199,9 +194,8 @@ def PrintAnalysisStageProgressInformation(self):
-    def _CalculateTotalDisplacement(self, node, old_total_displacement):
-        stage_displacement = node.GetSolutionStepValue(KratosMultiphysics.DISPLACEMENT)
-        total_displacement = old_total_displacement + stage_displacement
+    def _CalculateTotalDisplacement(self, node):
+        total_displacement = node.GetSolutionStepValue(KratosGeo.TOTAL_DISPLACEMENT) + node.GetSolutionStepValue(KratosGeo.INCREMENTAL_DISPLACEMENT)
```
We intend to move the implementation of the total displacement calculation to the C++ code (similar to how we calculate the incremental displacements). But that will be part of a separate PR.

markelov208 commented on applications/GeoMechanicsApplication/custom_workflows/solving_strategy_wrapper.hpp:
```diff
@@ -95,28 +95,10 @@ class SolvingStrategyWrapper : public StrategyWrapper
-                ++count;
-            }
+        for (auto& node : mrModelPart.Nodes()) {
+            node.GetSolutionStepValue(TOTAL_DISPLACEMENT) += node.GetSolutionStepValue(INCREMENTAL_DISPLACEMENT);
```
INCREMENTAL_DISPLACEMENT is much clearer than DISPLACEMENT. Does it mean that TOTAL_DISPLACEMENT is never cleaned any longer?

markelov208 replied:
I see that there is  akeyword in ProjectParameters file.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_workflows/test_solving_strategy_wrapper.cpp:
```diff
@@ -207,32 +207,6 @@ KRATOS_TEST_CASE_IN_SUITE(IncrementStepNumberFromStrategyWrapper, KratosGeoMecha
     KRATOS_EXPECT_EQ(wrapper.GetStepNumber(), 5);
 }
 
-KRATOS_TEST_CASE_IN_SUITE(SaveAndAccumulateTotalDisplacementField, KratosGeoMechanicsFastSuite)
```
This unit test had become redundant due to some simplifications.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_workflows/test_time_loop_executor.cpp:
```diff
@@ -319,13 +311,6 @@ KRATOS_TEST_CASE_IN_SUITE(ExpectRestoreCalledForTwoCycles, KratosGeoMechanicsFas
     KRATOS_EXPECT_TRUE(solver_strategy->IsRestoreCalled())
 }
 
-KRATOS_TEST_CASE_IN_SUITE(ExpectDisplacementFieldStoredForResetDisplacements, KratosGeoMechanicsFastSuiteWithoutKernel)
-{
-    auto solver_strategy = RunFixedCycleTimeLoop(1);
-
-    KRATOS_EXPECT_TRUE(solver_strategy->IsSaveFieldCalled())
-}
```
This unit test had become redundant due to some simplifications.

avdg81 commented on applications/GeoMechanicsApplication/tests/one_dimensional_consolidation/ProjectParameters_stage1.json:
```diff
@@ -23,13 +23,14 @@
         "nodal_smoothing":                    false,
         "block_builder":                      true,
-        "solution_type":                      "Quasi-Static",
+        "solution_type":                      "dynamic",
```
I have changed this `"solution_type"` to avoid the need to add another time integration scheme to the factory of the C++ workflow. Later on, when the improvements for the time integration schemes have been merged to `master`, we will modify this test to use "Quasi-static" in all stages.

avdg81 commented on applications/GeoMechanicsApplication/tests/settlement_workflow.py:
```diff
@@ -154,41 +154,19 @@ class KratosGeoMechanicsSettlementWorkflowCppRoute(KratosGeoMechanicsSettlementW
-        # components when it's destroyed. If no action is taken, any following tests will start to fail due to
-        # components not being registered. It seems that reloading the relevant Kratos applications overcomes this
-        # problem.
-        importlib.reload(KratosMultiphysics.LinearSolversApplication)
-        importlib.reload(KratosMultiphysics.StructuralMechanicsApplication)
-        importlib.reload(KratosMultiphysics.GeoMechanicsApplication)
```
Somehow, this is magically no longer needed... When these lines are kept, the following error is raised: `RuntimeError: Error: importing more than once the application : LinearSolversApplication`

markelov208 commented on applications/GeoMechanicsApplication/tests/one_dimensional_consolidation.py:
```diff
@@ -0,0 +1,146 @@
+        self.top_node_ids = [197, 198, 199, 200, 201]
+
+        self.end_times = [8640, 17280, 43200, 86400, 172800, 432000, 864000, 1728000, 4320000, 8640000]
+        self.t_vs = [0.01, 0.02, 0.05, 0.1, 0.2, 0.5, 1.0, 2.0, 5.0, 10.0]
```
This is the same time: first line is time in seconds and the second line is nondimensional time. I think a comment shall be here at least. The best is to calculate t_vs from end_times. `TT=C*T/( H * H)` H is available in the next line. The rest of data is in _check_degree_of_consolidation

avdg81 replied:
Agreed. The dimensionless times `self.t_vs` are now calculated based on `self.end_times`.

markelov208 commented on applications/GeoMechanicsApplication/tests/one_dimensional_consolidation.py:
```diff
@@ -0,0 +1,146 @@
+
+    def _check_relative_water_pressures_at_mid_column(self):
+        y_coordinates = self._get_y_coordinates_of_nodes(self.mid_column_node_ids)
+        stage_no = 2  # The first stage is not checked
```
Will the first stage break something? If not, then it is better to keep it for checking.

avdg81 replied:
When using the end time of stage 1 (i.e. $`t = 0`$), the calculation of the analytical solution will end up in an infinite loop, since `jt` equals 0.0 for each iteration. Note that Verruijt wrote _"This means that for large values
of time the series can be approximated by its first term"_, followed by the analytical formula that we've implemented. In other words, we shouldn't be using the given analytical formula when time values are small. Therefore, I think it's wise to exclude the first stage from checking.

markelov208 replied:
Thank you for the clarification. It seems this was the reason to introduce the max number of iterations.

markelov208 commented on applications/GeoMechanicsApplication/tests/one_dimensional_consolidation.py:
```diff
@@ -0,0 +1,146 @@
+        output_file_path = os.path.join(self.test_path, f"1D-Consolidationtest_stage{stage_no}.post.res")
+        reader = test_helper.GiDOutputFileReader()
+        output_data = reader.read_output_from(output_file_path)
+        return [-1.0 * u[1] for u in reader.nodal_values_at_time("TOTAL_DISPLACEMENT", time, output_data, node_ids)]
```
Does `-1.0` mean that the total displacement is defined from top to down like a force? The analytical solution needs the positive values, would it be better to use `deformation `instead of `vertical_displacement`?

avdg81 replied:
Downward (total) displacements are negative. Settlement values on the other hand (which have the same absolute value and direction) are positive. I'll rename the function and add a comment to make that clear.

markelov208 commented on applications/GeoMechanicsApplication/tests/one_dimensional_consolidation/ProjectParameters_stage1.json:
```diff
@@ -85,17 +86,16 @@
                         "skin_output":         false,
                         "plane_output":        [],
-                        "nodal_results":       ["DISPLACEMENT","WATER_PRESSURE"],
+                        "nodal_results":       ["DISPLACEMENT","TOTAL_DISPLACEMENT","WATER_PRESSURE"],
```
DISPLACEMENT is not used any more.

avdg81 replied:
Removed. Thanks for pointing this out.

markelov208 commented on applications/GeoMechanicsApplication/tests/one_dimensional_consolidation.py:
```diff
@@ -0,0 +1,146 @@
+    def test_1d_consolidation(self):
+        test_helper.run_stages(self.test_path, self.number_of_stages)
+
+        self._check_relative_water_pressures_at_mid_column()
```
Currently, relative water pressure is checked in the middle of the column when pore pressure was checked. Is it more sensitive check?

markelov208 replied:
Above you wrote that calculate_pore_pressure_1d_consolidation  would be removed later. Please could explain the reasons? Thank you.

avdg81 replied:
The only reason for using a selection of nodes (i.e. the mid-column ones) was to get a clearer plot, since then the y-coordinates change monotonically along the cross-section. But for checking the differences with the analytical solution, there is no need to use a node selection. So the test has been modified such that it checks the relative water pressure at all nodes.

And no, there was no good reason for keeping `calculate_pore_pressure_1d_consolidation`. Somehow, I thought that it was still used elsewhere, and I wanted to avoid scope creep. But now that I have checked it, it appears that the function is no longer used. So I have removed it. Good catch!

markelov208 reviewed (commented):
Hi Anne, many thanks for the clear implementation which is used for both python and cpp routes now. A use of the total displacement makes the code cleaner and more understandable. I have a few non-blocking comments.

avdg81 committed:
Processed review suggestions

- Removed a Python function that was no longer used.
- Check the relative water pressure values at all nodes rather than at the mid-column nodes only.

avdg81 committed:
Processed review suggestions

- Compute the dimensionless times `t_vs` using the end times of the stages.
- Use the symbol `h` to denote the sample height, to make it easier to check whether Verruijt's formulas have been implemented correctly.
- Renamed a member function and added a comment explaining why the sign of total displacement values is reversed.
- Removed unused output item "DISPLACEMENT" from the tests.

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/activate-1d-consolidation-test

markelov208 reviewed (approved):
Hi Anne, thank you very much for the changes. I think the PR is very ready to be merged.
